### PR TITLE
fix the fix that was supposed to fix #48. also added a unit test.

### LINKED ISSUE #48: ros/time.h fromSec implementation suffers from floating point inaccuracy, and stores invalid data
In some corner cases, floating point inaccuracy can lead to invalid data where nsec is calculated to be outside [0;1e9-1].
To illustrate the problem think of the following 

```
#include <cmath>
#include <boost/math/special_functions/round.hpp>
int main()
{
        double someint = 1031.0; // some integer
        double t = std::nextafter(someint, 0); // someint - epsilon
        std::cout << next << std::endl;    // prints 1031.000000
        // ros implementation of fromSec
        uint32_t sec = (uint32_t)floor(t);
        uint32_t nsec = (uint32_t)boost::math::round((t-sec) * 1e9);
        std::cout << "sec"  << sec  << std::endl;   // prints 1030 (!)
        std::cout << "nsec"<< nsec << std::endl;  // prints 1000000000 (!)
}
```

(compile with c++11)
Correct behavior would be sec: 1031 nsec: 0, of course.

I guess the rounding behavior is desired in general, so a remaining fix would be to catch this specific case. Alternatively, the order could be changed to multiply with 1e9 -> round -> split into secs/nsec.
The latter solution requires an appropriately long integer data type for the temporary.

### PR DESCRIPTION
**Pull request by pgao**
I made a unit test to test issue #48 and found that it failed. So I made a fix to the fix in #49 that was supposed to fix the issue that wasn't fixed, but now the test is fixed and the fix is fixed and I think the issue is fixed by the fix to the fix and yeah.

@televator

### CODE REVIEW

**Comment by televator** [comment]
Thanks, no idea how that got in there.

File: rostime/CMakeLists.txt [hunk 1]
```diff
@@ -1,6 +1,8 @@
 cmake_minimum_required(VERSION 2.8.3)
 project(rostime)
 
+set(CMAKE_CXX_FLAGS "-std=c++11 ${CMAKE_CXX_FLAGS}")
```
Comment (dirk-thomas) [inline rostime/CMakeLists.txt]: Can you please explain why C++11 is necessary here?

Comment (pgao) [inline rostime/CMakeLists.txt]: No particular reason, I just wanted to make sure the exact code in #48 compiled... mostly 4 da lulz. Is there a non-C++11 equivalent to std::nextafter?

Comment (dirk-thomas) [inline rostime/CMakeLists.txt]: If the flag is only required for the test then please set it only for that test target.

**Comment by pgao** [comment]
@dirk-thomas I've moved the C++11 into the test target. Let me know if there's anything else I can fix for the fix to the fix to the fix.

File: rostime/CMakeLists.txt [hunk 2]
```diff
@@ -38,6 +38,7 @@ if(CATKIN_ENABLE_TESTING)
   endif()
   catkin_add_gtest(${PROJECT_NAME}-test_time test/time.cpp)
   if(TARGET ${PROJECT_NAME}-test_time)
+    set(CMAKE_CXX_FLAGS "-std=c++11 ${CMAKE_CXX_FLAGS}")
```
Comment (dirk-thomas) [inline rostime/CMakeLists.txt]: This still changes the C++ flags globally and effects all target in this project. Please use something like `set_property(TARGET ${PROJECT_NAME}-test_time APPEND_STRING PROPERTY COMPILE_FLAGS "-std=c++11")` to set it specifically for this target.

**Comment by pgao** [comment]
@dirk-thomas look right?

**Comment by dirk-thomas** [comment]
Thank you for providing the test and fixing the problem.
